app/models/spree/address_decorator.rb

```diff
@@ -60,4 +60,13 @@ def update(attribute_names = @attributes.keys)
 
     super(attribute_names)
   end
+
+  def state_text
+    case Rails.configuration.state_text_attribute
+    when 'abbr'
+      state.try(:abbr)
+    when 'name'
+      state.try(:name) || state_name
+    end
+  end
```

Comment:
This is the original method we're overriding: https://github.com/spree/spree/blob/2-0-stable/core/app/models/spree/address.rb#L43

Reply:
This is good, but responsibility for determining how to display the name of the state feels like it should belong to the state rather than the address? It's just a thought, feel free to ignore it, but what about this:

````ruby
# state_decorator.rb
Spree::State.class_eval do
  ...
  # New method, not an override
  def display_text(fallback = nil)
    case Rails.configuration.state_text_attribute
    when 'abbr'
      abbr.try(:upcase)
    when 'name'
      try(:name) || fallback
    end
  end
  ...
end

# address_decorator.rb
Spree::Address.class_eval do
  ...
  def state_text
    state.try(:display_text, state_name)
  end
  ...
end
````

The best part about that is that then you can use `display_text` in the `StateSerializer` which removes the need for `RegistrationHelper#state_options_query `.

Another thing that I noticed is that there are no `Spree::Address` records in the Australian production database that have a non-empty `state_name`. I seems like if you provide a valid `state_id`, then `state_name` is just ignored?  Is this different for other countries? I am wondering whether the fallback to `address.state_name` is actually important?

Reply:
I've just realised that the above breaks the fallback to `state_name` when the state doesn't exist. To fix it you would need to interrogate `Rails.configuration.state_text_attribute` from `Spree::Address#state_text` which makes it messy. Ignore me, I think what you have is better...

Reply:
Thanks! Please approve the PR whenever you finish to review it.

Reply:
And no, `Address#state_name` is not used anymore is legacy stuff. I kept it for backward compatibility only